atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/lock/watch/LockWatchEteTest.java

```diff
@@ -0,0 +1,229 @@
+        lockWatcher.registerLockWatch(FULL_TABLE);
+        LockToken lockAfterToken = lock(lockAfterRegister);
+
+        TransactionHandler handler = startTransaction();
+
+        lockWatcher.unlock(lockBeforeToken);
+        lockWatcher.unlock(lockAfterToken);
+
+        assertThat(handler.getState(lockBeforeRegister)).isEqualTo(LockWatchInfo.State.LOCKED);
+        assertThat(handler.getState(lockAfterRegister)).isEqualTo(LockWatchInfo.State.LOCKED);
+    }
+
+    @Test
+    public void commitCanSeeLocksTakenOutAfterStartTransaction() {
+        LockDescriptor lockedBeforeRelockedAfterStart = getRandomRowDescriptor(TABLE);
+        LockDescriptor lockedAfterStart = getRandomRowDescriptor(TABLE);
+        LockDescriptor lockedAndUnlockedAfterStart = getRandomRowDescriptor(TABLE);
+
+        lockWatcher.registerLockWatch(FULL_TABLE);
+
+        LockToken lockedBeforeRelockedAfterToken = lock(lockedBeforeRelockedAfterStart);
+
+        TransactionHandler handler = startTransaction();
+
+        lockWatcher.unlock(lockedBeforeRelockedAfterToken);
+        lock(lockedBeforeRelockedAfterStart);
+        lock(lockedAfterStart);
+        LockToken lockedAndUnlockedAfterToken = lock(lockedAndUnlockedAfterStart);
+        lockWatcher.unlock(lockedAndUnlockedAfterToken);
+
+        assertThat(handler.commitAndGetLocks()).containsExactlyInAnyOrder(
+                lockedBeforeRelockedAfterStart,
+                lockedAfterStart,
+                lockedAndUnlockedAfterStart);
+    }
+
+    @Test
+    public void commitDoesNotSeeLocksTakenOutOnlyBeforeStartTransaction() {
+        LockDescriptor lockedBeforeStart = getRandomRowDescriptor(TABLE);
+        LockDescriptor lockedBeforeUnlockedAfterStart = getRandomRowDescriptor(TABLE);
+
+        lockWatcher.registerLockWatch(FULL_TABLE);
+        lock(lockedBeforeStart);
+        LockToken token = lock(lockedBeforeUnlockedAfterStart);
+
+        TransactionHandler handler = startTransaction();
+
+        lockWatcher.unlock(token);
+
+        assertThat(handler.commitAndGetLocks()).isEmpty();
+    }
+
+    @Test
+    public void filtersOutOurOwnLocks() {
+        lockWatcher.registerLockWatch(FULL_TABLE);
+        TransactionHandler handler = startTransaction();
+        LockDescriptor ourLockDescriptor = getRandomRowDescriptor(TABLE);
+        LockToken ourToken = lock(ourLockDescriptor);
+        assertThat(handler.commitAndGetLocksIgnoring(ourToken)).isEmpty();
+    }
+
+    @Test
+    public void doesNotFilerOutOurOwnLockIfLockedTwice() {
+        lockWatcher.registerLockWatch(FULL_TABLE);
+        TransactionHandler handler = startTransaction();
+        LockDescriptor ourLockDescriptor = getRandomRowDescriptor(TABLE);
+        LockToken token = lock(ourLockDescriptor);
+        lockWatcher.unlock(token);
+        LockToken ourToken = lock(ourLockDescriptor);
+        assertThat(handler.commitAndGetLocksIgnoring(ourToken)).containsExactly(ourLockDescriptor);
```

Comment:
discussed offline: This manifests in practice when our transaction reads and writes things falling under the same lock descriptor. We could otherwise miss a write-write conflict. Thanks for the test, it's a good one 👍